### PR TITLE
Making DeleteShard more resilient.

### PR DESCRIPTION
**Pull request by alainjobart**
If we can't find a ShardReplication object, fallback to reading all the
tablets in a cell (canonical tablet records) to find potential matches.

Also adding a unit test for that logic.

### CODE REVIEW

File: go/vt/wrangler/shard.go [hunk 1]
```diff
@@ -121,32 +123,85 @@ func (wr *Wrangler) DeleteShard(ctx context.Context, keyspace, shard string, rec
...
+				return fmt.Errorf("shard %v/%v still has %v tablets in cell %v; use -recursive or remove them manually", keyspace, shard, len(tabletMap), cell)
+			}
+
+			wr.Logger().Infof("Deleting all tablets in shard %v/%v cell %v", keyspace, shard, cell)
+			for tabletAlias := range tabletMap {
+				// We don't care about scrapping or updating the replication graph,
+				// because we're about to delete the entire replication graph.
+				wr.Logger().Infof("Deleting tablet %v", topoproto.TabletAliasString(&tabletAlias))
+				if err := wr.TopoServer().DeleteTablet(ctx, &tabletAlias); err != nil && err != topo.ErrNoNode {
+					// Unlike the errors below in non-recursive steps, we don't want to
```
Comment (pivanof) [inline go/vt/wrangler/shard.go]: Which "errors below" is this talking about?

File: go/vt/wrangler/testlib/fake_tablet.go [hunk 2]
```diff
@@ -127,7 +127,7 @@ func NewFakeTablet(t *testing.T, wr *wrangler.Wrangler, cell string, uid uint32,
 	delete(tablet.PortMap, "start_http_server")
 	_, force := tablet.PortMap["force_init"]
 	delete(tablet.PortMap, "force_init")
-	if err := wr.InitTablet(context.Background(), tablet, force, true, false); err != nil {
+	if err := wr.InitTablet(context.Background(), tablet, force, true /*createShardAndKeyspace*/, false /*allowUpdate*/); err != nil {
```
Comment (pivanof) [inline go/vt/wrangler/testlib/fake_tablet.go]: How is this related?

File: go/vt/wrangler/testlib/shard_test.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	// Now try to delete the shard without ignore_if_serving flag,
+	// should fail.
+	if err := vp.Run([]string{
+		"DeleteShard",
+		master.Tablet.Keyspace + "/" + master.Tablet.Shard,
+	}); err == nil || !strings.Contains(err.Error(), "is still serving, cannot delete it") {
+		t.Fatalf("DeleteShard() returned wrong error: %v", err)
+	}
+
+	// Now try to delete the shard without recursive flag, should fail.
```
Comment (pivanof) [inline go/vt/wrangler/testlib/shard_test.go]: This comment is confusing. "Now try ... without recursive flag" suggests that we've already tried with recursive flag, but I don't see that above.

File: go/vt/vtctl/vtctl.go [hunk 4]
```diff
@@ -1468,6 +1468,7 @@ func commandRemoveShardCell(ctx context.Context, wr *wrangler.Wrangler, subFlags
 
 func commandDeleteShard(ctx context.Context, wr *wrangler.Wrangler, subFlags *flag.FlagSet, args []string) error {
 	recursive := subFlags.Bool("recursive", false, "Also delete all tablets belonging to the shard.")
+	ignoreIfServing := subFlags.Bool("ignore_if_serving", false, "Remove the shard even if it serving. Use with caution.")
```
Comment (alainjobart) [inline go/vt/vtctl/vtctl.go]: Agreed, will use 'even if serving', better.

File: go/vt/wrangler/shard.go (hunk 1 above)
Comment (alainjobart) [inline go/vt/wrangler/shard.go]: Old cut&paste comment. Will fix.

File: go/vt/wrangler/testlib/shard_test.go (hunk 3 above)
Comment (alainjobart) [inline go/vt/wrangler/testlib/shard_test.go]: Updated the comments.

**Comment by alainjobart** [comment]
Uploaded fix, PTAL.

File: go/vt/vtctl/vtctl.go [hunk 5]
```diff
@@ -1468,6 +1468,7 @@ func commandRemoveShardCell(ctx context.Context, wr *wrangler.Wrangler, subFlags
 
 func commandDeleteShard(ctx context.Context, wr *wrangler.Wrangler, subFlags *flag.FlagSet, args []string) error {
 	recursive := subFlags.Bool("recursive", false, "Also delete all tablets belonging to the shard.")
+	evenIfServing := subFlags.Bool("even_if_serving", false, "Remove the shard even if it serving. Use with caution.")
```
Comment (pivanof) [inline go/vt/vtctl/vtctl.go]: s/it serving/it is serving/

File: go/vt/wrangler/keyspace.go [hunk 6]
```diff
@@ -850,7 +850,7 @@ func (wr *Wrangler) DeleteKeyspace(ctx context.Context, keyspace string, recursi
 		wr.Logger().Infof("Deleting all shards (and their tablets) in keyspace %v", keyspace)
 		for _, shard := range shards {
 			wr.Logger().Infof("Recursively deleting shard %v/%v", keyspace, shard)
-			if err := wr.DeleteShard(ctx, keyspace, shard, recursive); err != nil && err != topo.ErrNoNode {
+			if err := wr.DeleteShard(ctx, keyspace, shard, true /*recursive*/, true /*ignoreIfServing*/); err != nil && err != topo.ErrNoNode {
```
Comment (pivanof) [inline go/vt/wrangler/keyspace.go]: s/ignoreIfServing/evenIfServing/

And it's better to have spaces around parameter names (i.e. "/* recursive */").

File: go/vt/wrangler/testlib/shard_test.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+	slave := NewFakeTablet(t, wr, "cell1", 1, topodatapb.TabletType_REPLICA, db)
+	remoteSlave := NewFakeTablet(t, wr, "cell2", 2, topodatapb.TabletType_REPLICA, db)
+
+	// Delete the ShardReplication record in cell2
+	if err := ts.DeleteShardReplication(ctx, "cell2", remoteSlave.Tablet.Keyspace, remoteSlave.Tablet.Shard); err != nil {
+		t.Fatalf("DeleteShardReplication failed: %v", err)
+	}
+
+	// Now try to delete the shard without even_if_serving or
+	// resursive flag, should fail on serving check first.
```
Comment (pivanof) [inline go/vt/wrangler/testlib/shard_test.go]: s/resursive/recursive/

**Comment by alainjobart** [comment]
@pivanof Addressed test failures, PTAL.

File: go/vt/wrangler/testlib/fake_tablet.go (hunk 2 above)
Comment (pivanof) [inline go/vt/wrangler/testlib/fake_tablet.go]: At least it would be nice to follow the same rule and put spaces around the parameter names ("/* allowUpdate */").

Comment (alainjobart) [inline go/vt/wrangler/testlib/fake_tablet.go]: Done. Also squashed.
